### DISCUSSION TITLE
How to use colored text in Quarto figure captions

### DISCUSSION THREAD

**Discussion by friendly** [quarto-dev/quarto-cli#10299] [Q&A]
### Description

In figure captions, it is often useful to say refer to colored objects in graphs by printing
the words in their colors: e.g., 
"The blue points show the data values. The regression line is shown in green, while the loess smooth is shown in red"

In the working example below, I can color words in text using an inline R expression, `r colorize("red")`, which works both for HTML and PDF output. However, this does not work in a figure caption.  I realize that this may involve `pandoc`, but I'd like to find a solution that works.  Can anyone help?

My working example:

````qmd
---
title: "How to use colored text in Quarto figure captions"
format:
  html:
    toc: true
  pdf:
    toc: true
---
  
In figure captions, it is often useful to say refer to colored objects by printing
the words in their colors.
How can I do this with output in HTML or PDF?

Define a function to colorize text, that can be used in HTML or PDF output

```{r colorize}
# colorize text: use inline as `r colorize(text, color)` to print `text` in a given `color`
# can also be used to color a color name, as in r colorize("red")`
colorize <- function(text, color) {
  if (missing(color)) color <- text
  if (knitr::is_latex_output()) {
    sprintf("\\textcolor{%s}{%s}", color, text)
  } else if (knitr::is_html_output()) {
    sprintf("<span style='color: %s;'>%s</span>", color, text)
  } else text
}
```

I can print text in `r colorize("red")`, `r colorize("blue")`, `r colorize("green")`, ... `

But not in a figure caption.

This works without color text in the caption:
````

```{r}
#| label: fig-test
#| out-width: "70%"
#| fig-cap: Some points should be red, some blue, some green
cols <- c("red", "blue", "green")
plot(1:10, pch=16, cex = 3, col = cols)
```

![image](https://github.com/user-attachments/assets/a95d9de8-1634-4e2f-8e5d-9e2ed62758c1)

This doesn't work:

```{r}
#| label: fig-test2
#| out-width: "70%"
#| fig-cap: Some points are `r colorize("red")`, some are `r colorize("blue")`, some are `r colorize("green")`
cols <- c("red", "blue", "green")
plot(1:10, pch=16, cex = 3, col = cols)
```

![image](https://github.com/user-attachments/assets/dd0da060-ccdd-4081-9e3f-30cdbe23275b)

**Comment by friendly** [comment]
My working example can be found at: https://raw.githubusercontent.com/friendly/Vis-MLM-book/master/test/minimal-example.qmd

**Comment by mcanouil** [comment]
The inline R inside code cell is never going to work in Quarto or even in RMarkdown.

You need markdown (or use the knitr specific feature with `!expr` which I don't recommend especially if markdown works).

An helper could be to use the following extension: https://github.com/mcanouil/quarto-highlight-text


```txt
#| fig-cap: 'This is [Red]{colour="#b22222"}.'
```

The above is not tested.


The `!expr` would be something like:

```txt
#| fig-cap: !expr sprintf('This is %s.', colorize('red'))
```

**Comment by friendly** [comment]
Thanks so much @mcanouil for pointing me to `!expr`. 
`quarto-highlight-text` is interesting, but seems awkward to use, at least to me.

For writing figure captions, `paste()` is easier than `sprintf()`:


```{r}
#| label: fig-test3
#| out-width: "70%"
#| fig-cap: !expr paste("Some points are", colorize('red'), "some are", colorize('blue'), "some are", colorize("green"))
cols <- c("red", "blue", "green")
plot(1:10, pch=16, cex = 3, col = cols)
```

But there is an even better solution that makes such captions easier to write, by defining global variables with color names and using them with `glue::glue()`

```{r}
red <- colorize('red')
blue <- colorize('blue')
green <- colorize("green")
library(glue)
```

Then this works:
```{r}
#| label: fig-test4
#| out-width: "70%"
#| fig-cap: !expr glue("Some points are ", {red}, " some are ", {blue}, "some are ", {green})
plot(1:10, pch=16, cex = 3, col = cols)
```

Or just with `paste()`:

```
#| fig-cap: !expr paste("Some points are", red, "some are", blue, "some are", green)
```

**Reply by mcanouil** [reply]
My point was not to compare the various ways to concatenate strings. I simply picked one to show an example writable on a mobile.

Note that !expr is a very knitr specific feature that "works" in Quarto but not really intentionally. So you should not use it as your default strategy.

About the syntax, that's the standard Pandoc/Quarto/RMarkdowb syntax called span.
`[text in red]{style="color: red;"}` would work but only for HTML. My extension does this for several formats and also allows to set background colour.
It relies on nothing specific such as `!expr` and does not need any engines/computations. It's "native".

**Reply by friendly** [reply]
I need a solution that works for HTML and PDF, as does my `colorize()` function. I'll take another look at the highlight extension.

This is for a book project, where I expect to use this often, so ideally I want something that won't break and is not onerous to write.

**Reply by friendly** [reply]
One drawback of the highlight extension seems to be the need to specify colors in #rrggbb hex notation.

**Reply by mcanouil** [reply]
I would not call it a drawback especially if you intend to use colours in caption as a legend for your figure or if you are using multiple formats.

**Comment by friendly** [comment]
My complete example is now at: https://raw.githubusercontent.com/friendly/Vis-MLM-book/master/test/caption-colors.qmd

**Comment by cscheid** [comment]
> This doesn't work:
> 
> ```
> #| label: fig-test2
> #| out-width: "70%"
> #| fig-cap: Some points are `r colorize("red")`, some are `r colorize("blue")`, some are `r colorize("green")`
> cols <- c("red", "blue", "green")
> plot(1:10, pch=16, cex = 3, col = cols)
> ```

Here's a general strategy. Every time you feel tempted to reach out for inline expressions and `!expr` in a code cell, you should consider whether the div syntax for floats could work for you. Concretely, in this case:

<table><tr><th>qmd</th><th>result</th></tr>
<tr>
<td>

````
---
title: issue-10299
---

```{r}
#| echo: false
#| output: false
colorize <- function(text, color) {
  if (missing(color)) color <- text
  if (knitr::is_latex_output()) {
    sprintf("\\textcolor{%s}{%s}", color, text)
  } else if (knitr::is_html_output()) {
    sprintf("<span style='color: %s;'>%s</span>", color, text)
  } else text
}
```

::: {#fig-test}

```{r}
#| out-width: "70%"
cols <- c("red", "blue", "green")
plot(1:10, pch=16, cex = 3, col = cols)
```

Some points are `r colorize("red")`,
some are `r colorize("blue")`,
some are `r colorize("green")`.

:::

See @fig-test.

````

</td><td>

<img width="1082" alt="image" src="https://github.com/user-attachments/assets/c6574567-9c14-41dd-8598-c2d9aba40efb">

</td></tr></table>
